src/lavinmq/mqtt/client.cr

```diff
@@ -0,0 +1,261 @@
+require "openssl"
+require "socket"
+require "../client"
+require "../error"
+require "../rough_time"
+require "./session"
+require "./protocol"
+
+module LavinMQ
+  module MQTT
+    class Client < LavinMQ::Client
+      include Stats
+      include SortableJSON
+
+      getter channels, log, name, user, client_id, socket, remote_address, connection_info
+      getter? clean_session
+      @connected_at = RoughTime.unix_ms
+      @channels = Hash(UInt16, Client::Channel).new
+      @session : MQTT::Session?
+      rate_stats({"send_oct", "recv_oct"})
+      Log = LavinMQ::Log.for "mqtt.client"
+
+      def vhost
+        @broker.vhost
+      end
+
+      def initialize(@socket : ::IO,
+                     @connection_info : ConnectionInfo,
+                     @user : User,
+                     @broker : MQTT::Broker,
+                     @client_id : String,
+                     @clean_session : Bool = false,
+                     @keepalive : UInt16 = 30,
+                     @will : MQTT::Will? = nil)
+        @io = MQTT::IO.new(@socket)
+        @lock = Mutex.new
+        @waitgroup = WaitGroup.new(1)
+        @remote_address = @connection_info.src
+        @local_address = @connection_info.dst
+        @name = "#{@remote_address} -> #{@local_address}"
+        @metadata = ::Log::Metadata.new(nil, {vhost: @broker.vhost.name, address: @remote_address.to_s, client_id: client_id})
+        @log = Logger.new(Log, @metadata)
+        @broker.vhost.add_connection(self)
+        @log.info { "Connection established for user=#{@user.name}" }
+        spawn read_loop
+      end
+
+      def client_name
+        "mqtt-client"
+      end
+
+      private def read_loop
+        socket = @socket
+        if socket.responds_to?(:"read_timeout=")
+          socket.read_timeout = (@keepalive * 1.5).seconds
+        end
+        loop do
+          @log.trace { "waiting for packet" }
+          packet = read_and_handle_packet
+          # The disconnect packet has been handled and the socket has been closed.
+          # If we dont breakt the loop here we'll get a IO/Error on next read.
+          if packet.is_a?(MQTT::Disconnect)
+            @log.debug { "Recieved disconnect" }
+            break
+          end
+        end
+      rescue ex : ::MQTT::Protocol::Error::PacketDecode
+        @log.warn(exception: ex) { "Packet decode error" }
+        publish_will
+      rescue ex : MQTT::Error::Connect
```

Comment:
connect errors at this stage, is that possible? or i just guessed that Connect errors was releated to connection negotiation, but maybe it can be raised for other cases too?

Reply:
no, i think this is a rest from early day implementation